mono/metadata/threads.c

```diff
@@ -5065,15 +5088,67 @@ threads_add_joinable_thread_nolock (gpointer tid)
 }
 #endif
 
+static void
+threads_add_pending_joinable_runtime_thread (gpointer thread_info)
+{
+	g_assert (thread_info);
+	MonoThreadInfo *mono_thread_info = (MonoThreadInfo*)thread_info;
+
+	if (mono_thread_info->runtime_thread) {
+		gint32 old_joinable_state = mono_atomic_load_i32 (&mono_thread_info->thread_joinable_state);
```

Comment:
Shouldn't this be a CAS loop?
CAS is allowed to optimistically fail and return the old value.
Beyond that, what could cause this CAS to fail?

Reply:
So how do we detect that the CAS operation was successful using mono's CAS primitives if we can't make a difference between spurious failures and successful update (both returning the old value) as a result of the CAS operation? In the current design I need to know that the calling thread either was successfully updating the value or in case of failures (concurrent writes or spurious failure) that it failed to do so, then I can correctly reevaluate the value and potential retry (in case of a spurious failure).

In C++ you have the compare_exchange_strong or compare_exchange_weak, both returning a bool indicating success/failure, where failure is either due to concurrent writes and/or spurious failure. 

The above code needs the characteristics of compare_exchange_strong (or compare_exchange_weak in a loop revalidating the state) since it needs to know if it was the one actually doing the update or not.

I put in the CAS operations around thread_joinable_state as a safety since I didn't have full overview of code calling into mono_thread_detach_internal and mono_threads_add_joinable_runtime_thread and if those could in theory be called in parallell for the same mono_thread_info. If we could conclude that it would never happen, I could drop the CAS update of the fields. Doing the update using CAS was just some extra safety and I hardly expect them to actually fail.